src/server/pfs/s3/bucket.go

```diff
@@ -0,0 +1,359 @@
+	Owner  User   `xml:"Owner"`
+}
+
+func newCommonPrefixes(dir string) CommonPrefixes {
+	return CommonPrefixes{
+		Prefix: fmt.Sprintf("%s/", dir),
+		Owner:  defaultUser,
+	}
+}
+
+type bucketHandler struct {
+	pc *client.APIClient
+}
+
+func newBucketHandler(pc *client.APIClient) bucketHandler {
+	return bucketHandler{pc: pc}
+}
+
+func (h bucketHandler) location(w http.ResponseWriter, r *http.Request) {
+	repo, branch := bucketArgs(w, r)
+	
+	_, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/xml")
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte(locationSource))
+}
+
+func (h bucketHandler) get(w http.ResponseWriter, r *http.Request) {
+	repo, branch := bucketArgs(w, r)
+
+	// ensure the branch exists and has a head
+	branchInfo, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+
+	maxKeys := defaultMaxKeys
+	maxKeysStr := r.FormValue("max-keys")
+	if maxKeysStr != "" {
+		i, err := strconv.Atoi(maxKeysStr)
+		if err != nil || i < 0 || i > defaultMaxKeys {
+			invalidArgument(w, r)
+			return
+		}
+		maxKeys = i	
+	}
+
+	delimiter := r.FormValue("delimiter")
+	if delimiter != "" && delimiter != "/" {
+		invalidDelimiterError(w, r)
+		return
+	}
+
+	result := &ListBucketResult{
+		Name:        repo,
+		Prefix:      r.FormValue("prefix"),
+		Marker:      r.FormValue("marker"),
+		Delimiter:   delimiter,
+		MaxKeys:     maxKeys,
+		IsTruncated: false,
+	}
+
+	if branchInfo.Head == nil {
+		// if there's no head commit, just print an empty list of files
+		writeXML(w, r, http.StatusOK, result)
+	} else if delimiter == "" {
+		h.listRecursive(w, r, result, branch)
+	} else {
+		h.list(w, r, result, branch)
+	}
+}
+
+func (h bucketHandler) listRecursive(w http.ResponseWriter, r *http.Request, result *ListBucketResult, branch string) {
+	err := h.pc.Walk(result.Name, branch, filepath.Dir(result.Prefix), func(fileInfo *pfs.FileInfo) error {
+		origFilePath := fileInfo.File.Path
+		fileInfo = updateFileInfo(branch, result.Marker, result.Prefix, fileInfo)
+		if fileInfo == nil || fileInfo.FileType == pfs.FileType_DIR {
+			return nil
+		}
+		if result.isFull() {
+			if result.MaxKeys > 0 {
+				result.IsTruncated = true
+			}
+			return errutil.ErrBreak
+		}
+
+		meta, err := getMeta(h.pc, result.Name, branch, origFilePath)
+		if err != nil {
+			return err
+		}
+		etag := ""
+		if meta != nil {
+			etag = meta.MD5
+		}
+		
+		contents, err := newContents(fileInfo, etag)
+		if err != nil {
+			return err
+		}
+		result.Contents = append(result.Contents, contents)
+		
+		return nil
+	})
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	setNextMarker(result)
+	writeXML(w, r, http.StatusOK, result)
+}
+
+func (h bucketHandler) list(w http.ResponseWriter, r *http.Request, result *ListBucketResult, branch string) {
+	pattern := fmt.Sprintf("%s*", glob.QuoteMeta(result.Prefix))
```

Comment:
Clarification on this: why is this `%s*` and not `%s**`? My understanding of s3's path listing was that it doesn't really have a concept of directories so when you do a list with a prefix you get back everything underneath that path including nested directories. Am I wrong about that? Or am I wrong about what this is doing? Or am I just wrong about everything?

Reply:
Oh, I guess that's what `listRecursive` does above, that makes sense.

Reply:
I wasn't aware of `%s**`, and it doesn't seem to be supported by go's stdlib [glob matcher](https://golang.org/pkg/path/filepath/#Match). Is that an extension PFS supports? Does it recurse into subdirectories? If so, I could probably simplify this by nixing `listRecursive`.

Reply:
Yeah, we use a different glob library with some extensions, that being the main one.

Reply:
Done